src/windows/osd.c

```diff
@@ -402,64 +402,109 @@ ssize_t sendmsg(SOCKET sd, struct msghdr *msg, int flags)
 
-	assert(ifap);
-
-	if (res == ERROR_INSUFFICIENT_BUFFER) {
-		iptbl = malloc(ips);
-		if (!iptbl)
-			goto failed_no_mem;
-		res = GetIpAddrTable(iptbl, &ips, 0);
-		if (res != NO_ERROR)
-			goto failed_get_addr;
-	} else if (res != NO_ERROR) {
+	res = GetAdaptersAddresses(AF_UNSPEC, GAA_FLAG_INCLUDE_PREFIX, NULL, NULL, &size);
+	if (res != ERROR_BUFFER_OVERFLOW) {
+		goto failed_no_mem;
+	}
+	adapter_addresses = (PIP_ADAPTER_ADDRESSES)malloc(size);
+	res = GetAdaptersAddresses(AF_UNSPEC, GAA_FLAG_INCLUDE_PREFIX, NULL, adapter_addresses, &size);
+	if (res != ERROR_SUCCESS) {
+		free(adapter_addresses);
 		goto failed;
 	}
 
-	for (i = 0; i < iptbl->dwNumEntries; i++) {
-		if (iptbl->table[i].dwAddr && iptbl->table[i].dwAddr != ntohl(INADDR_LOOPBACK)) {
-			struct ifaddrs *fa = calloc(sizeof(*fa), 1);
-			if (!fa)
-				goto failed_cant_allocate;
-			fa->ifa_flags = IFF_UP;
-			fa->ifa_addr = (struct sockaddr *)&fa->in_addr;
-			fa->ifa_netmask = (struct sockaddr *)&fa->in_netmask;
-			fa->ifa_name = fa->ad_name;
-
-			fa->in_addr.sin_family = fa->in_netmask.sin_family = AF_INET;
-			fa->in_addr.sin_addr.s_addr = iptbl->table[i].dwAddr;
-			fa->in_netmask.sin_addr.s_addr = iptbl->table[i].dwMask;
-			/* on Windows there is no Unix-like interface names,
-			   so, let's generate fake names */
-			sprintf_s(fa->ad_name, sizeof(fa->ad_name), "eth%d", i);
-
-			fa->ifa_next = head;
-			head = fa;
+	sort_list_of_adapters(adapter_addresses);
+
+	for (aa = adapter_addresses; aa != NULL; aa = aa->Next) {
+		if (aa->OperStatus == 1) {
+			for (ua = aa->FirstUnicastAddress; ua != NULL; ua = ua->Next) {
+				SOCKADDR *pSockAddr = ua->Address.lpSockaddr;
+				struct ifaddrs *fa = calloc(sizeof(*fa), 1);
+				if (!fa)
+					goto failed_cant_allocate;
+
+				fa->ifa_flags = IFF_UP;
+				if (aa->IfType == IF_TYPE_SOFTWARE_LOOPBACK) {
+					fa->ifa_flags |= IFF_LOOPBACK;
+				}
+				fa->ifa_addr = (struct sockaddr *)&fa->in_addr;
+				fa->ifa_netmask = (struct sockaddr *)&fa->in_netmask;
+				fa->ifa_name = fa->ad_name;
+
+				fa->in_addr.sin_family = fa->in_netmask.sin_family = pSockAddr->sa_family;
+				if (pSockAddr->sa_family == AF_INET) {
+					ULONG subnet = 0;
+					PULONG mask = &subnet;
+					struct sockaddr_in *pInAddr = (struct sockaddr_in*)pSockAddr;
+					fa->in_addr.sin_addr = pInAddr->sin_addr;
+					if (ConvertLengthToIpv4Mask(ua->OnLinkPrefixLength, mask) == NO_ERROR) {
+						fa->in_netmask.sin_addr.S_un.S_addr = mask;
+					}
+					else {
+						goto failed;
+					}
+				} else if (pSockAddr->sa_family == AF_INET6) {
+					continue;
+				} else {
+					continue;
+				}
+				sprintf_s(fa->ad_name, sizeof(fa->ad_name), "eth%d", i++);
+				fa->ifa_next = head;
+				head = fa;
+			}
 		}
 	}
-
-	if (iptbl != &_iptbl)
-		free(iptbl);
+	free(adapter_addresses);
 	ret = 0;
 	if (ifap)
 		*ifap = head;
 complete:
 	return ret;
-
 failed_cant_allocate:
-	if(head)
+	if (head)
 		freeifaddrs(head);
-failed_get_addr:
-	free(iptbl);
```

Comment:
We need to free(adapter_addresses) somewhere down here

Reply:
@shefty Done. We are doing it in `failed` path now.